cmd/goupnpdcpgen/metadata.go

```diff
@@ -1,69 +1,116 @@
-		Hacks: []DCPHackFn{
-			func(dcp *DCP) error {
-				missingURN := "urn:schemas-upnp-org:service:WANIPv6FirewallControl:1"
-				if _, ok := dcp.ServiceTypes[missingURN]; ok {
-					return nil
-				}
-				urnParts, err := extractURNParts(missingURN, serviceURNPrefix)
-				if err != nil {
-					return err
-				}
-				dcp.ServiceTypes[missingURN] = urnParts
-				return nil
-			}, totalBytesHack,
+		Src: upnpdotorg{
+			DocURL:     "http://upnp.org/specs/gw/UPnP-gw-InternetGatewayDevice-v2-Device.pdf",
+			XMLSpecURL: "http://upnp.org/specs/gw/UPnP-gw-IGD-Testfiles-20110224.zip",
+			Hacks: []DCPHackFn{
+				fixMissingURN("urn:schemas-upnp-org:service:WANIPv6FirewallControl:1"),
+				totalBytesHack("urn:schemas-upnp-org:service:WANCommonInterfaceConfig:1"),
+			},
 		},
 	},
 	{
 		Name:         "av1",
 		OfficialName: "MediaServer v1 and MediaRenderer v1",
-		DocURL:       "http://upnp.org/specs/av/av1/",
-		XMLSpecURL:   "http://upnp.org/specs/av/UPnP-av-TestFiles-20070927.zip",
+		Src: upnpdotorg{
+			DocURL:     "http://upnp.org/specs/av/av1/",
+			XMLSpecURL: "http://upnp.org/specs/av/UPnP-av-TestFiles-20070927.zip",
+		},
+	},
+	{
+		Name:         "ocf/internetgateway1",
+		OfficialName: "Internet Gateway Device v1 - Open Connectivity Foundation",
+		Src: openconnectivitydotorg{
+			SpecsURL:       ocfSpecsURL,
+			DocPath:        "*/DeviceProtection_1/UPnP-gw-*v1*.pdf",
+			XMLSpecZipPath: "*/DeviceProtection_1/UPnP-gw-IGD-TestFiles-*.zip",
+			XMLServicePath: []string{"*/service/*1.xml"},
+			XMLDevicePath:  []string{"*/device/*1.xml"},
+			Hacks: []DCPHackFn{
+				fixMissingURN("urn:schemas-upnp-org:service:DeviceProtection:1"),
+				fixMissingURN("urn:schemas-upnp-org:service:WANIPv6FirewallControl:1"),
+				totalBytesHack(""),
+			},
+		},
+	},
+	{
+		Name:         "ocf/internetgateway2",
+		OfficialName: "Internet Gateway Device v2 - Open Connectivity Foundation",
+		Src: openconnectivitydotorg{
+			SpecsURL:       ocfSpecsURL,
+			DocPath:        "*/Internet Gateway_2/UPnP-gw-*.pdf",
+			XMLSpecZipPath: "*/Internet Gateway_2/UPnP-gw-IGD-TestFiles-*.zip",
+			XMLServicePath: []string{"*/service/*1.xml", "*/service/*2.xml"},
+			XMLDevicePath:  []string{"*/device/*1.xml", "*/device/*2.xml"},
+			Hacks: []DCPHackFn{
+				fixMissingURN("urn:schemas-upnp-org:service:DeviceProtection:1"),
+				totalBytesHack(""),
+			},
+		},
 	},
 }
 
-func totalBytesHack(dcp *DCP) error {
-	for _, service := range dcp.Services {
-		if service.URN == "urn:schemas-upnp-org:service:WANCommonInterfaceConfig:1" {
-			variables := service.SCPD.StateVariables
-			for key, variable := range variables {
-				varName := variable.Name
-				if varName == "TotalBytesSent" || varName == "TotalBytesReceived" {
-					// Fix size of total bytes which is by default ui4 or maximum 4 GiB.
-					variable.DataType.Name = "ui8"
-					variables[key] = variable
+func totalBytesHack(missingURN string) func(dcp *DCP) error {
```

Comment:
This function should not take a string, I think it's trying to be too consistent with `fixMissingURN`. However where `totalBytesHack("")` appears, it should simply be `totalBytesHack`, and this function signature should simply be: `func totalBytesHack(dcp *DCP) error {`

Reply:
i was concerned it breaks this condition

https://github.com/huin/goupnp/blob/master/cmd/goupnpdcpgen/metadata.go#L51

I am guessing you meant it is ok to apply this hack on all services.

Reply:
Ah fair. But in that case it should presumably be taking the URN(s) as an argument rather than an empty string.